src/autoscaling/snapshotter.js

```diff
@@ -0,0 +1,170 @@
+import log from 'apify-shared/log';
+import { checkParamOrThrow } from 'apify-client/build/utils';
+import { getMemoryInfo, isAtHome } from '../utils';
+import { ACTOR_EVENT_NAMES } from '../constants';
+import events from '../events';
+
+const DEFAULT_OPTIONS = {
+    eventLoopSnapshotIntervalSecs: 0.5,
+    memorySnapshotIntervalSecs: 1,
+    snapshotHistorySecs: 60,
+    maxBlockedRatio: 0.1,
+    minFreeMemoryRatio: 0.7,
+};
+
+/**
+ * Creates snapshots of system resources at given intervals.
+ * Provides an interface to read captured snapshots.
+ *
+ * @param options
+ */
+export default class Snapshotter {
+    constructor(options = {}) {
+        const {
+            eventLoopSnapshotIntervalSecs,
+            memorySnapshotIntervalSecs,
+            snapshotHistorySecs,
+            maxBlockedRatio,
+            minFreeMemoryRatio,
+            maxMemoryMbytes,
+        } = _.defaults(options, DEFAULT_OPTIONS);
+
+        checkParamOrThrow(eventLoopSnapshotIntervalSecs, 'options.eventLoopSnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(memorySnapshotIntervalSecs, 'options.memorySnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(snapshotHistorySecs, 'options.snapshotHistorySecs', 'Number');
+        checkParamOrThrow(maxBlockedRatio, 'options.maxBlockedRatio', 'Number');
+        checkParamOrThrow(minFreeMemoryRatio, 'options.minFreeMemoryRatio', 'Number');
+        checkParamOrThrow(maxMemoryMbytes, 'options.maxMemoryMbytes', 'Maybe Number');
+
+        this.eventLoopSnapshotIntervalMillis = eventLoopSnapshotIntervalSecs * 1000;
+        this.memorySnapshotIntervalMillis = memorySnapshotIntervalSecs * 1000;
+        this.snapshotHistoryMillis = snapshotHistorySecs * 1000;
+        this.maxBlockedMillis = 1000 * eventLoopSnapshotIntervalSecs * maxBlockedRatio;
+        this.minFreeMemoryRatio = minFreeMemoryRatio;
+        if (maxMemoryMbytes) this.maxMemoryBytes = maxMemoryMbytes * 1024 * 1024;
+
+        this.cpuSnapshots = [];
+        this.eventLoopSnapshots = [];
+        this.memorySnapshots = [];
+    }
+
+    async start() {
+        // Ensure max memory is correctly computed.
+        if (!this.maxMemoryBytes) {
+            const { totalBytes } = await getMemoryInfo();
+            if (isAtHome()) this.maxMemoryBytes = totalBytes;
+            else this.maxMemoryBytes = Math.ceil(totalBytes / 4);
+        }
+        // Add dummy snapshot to compare the first with.
+        this.eventLoopSnapshots.push({
+            createdAt: new Date(),
+            isOverloaded: false,
+            exceededMillis: 0,
+        });
+
+        // Start snapshotting.
+        this.eventLoopInterval = betterSetInterval(this._snapshotEventLoop.bind(this), this.eventLoopSnapshotIntervalMillis);
+        this.memoryInterval = betterSetInterval(this._snapshotMemory.bind(this), this.memorySnapshotIntervalMillis);
+        if (isAtHome()) events.on(ACTOR_EVENT_NAMES.CPU_INFO, this._snapshotCpu.bind(this));
```

Comment:
You don't need isAtHome here. 
It won't simply trigger when there are no events ....

Reply:
I know, it just feels right for the code to document it somehow. That the events actually won't fire unless the run is at Platform. It has no performance impact.